framework/include/controls/ControlInterface.h

```diff
@@ -0,0 +1,219 @@
+/****************************************************************/
+
+#ifndef CONTROLINTERFACE_H
+#define CONTROLINTERFACE_H
+
+// MOOSE includes
+#include "InputParameters.h"
+#include "InputParameterWarehouse.h"
+#include "MooseUtils.h"
+
+// Forward declarations
+class ControlInterface;
+
+template<>
+InputParameters validParams<ControlInterface>();
+
+/**
+ * An interface class for accessing writable references to InputParameters
+ *
+ * This class exists to provided parameter access to the Control base class
+ * as well as the RealParameterReporter postprocessor, no other objects
+ * should use this class. Hence, the private constructor
+ */
+class ControlInterface
+{
+protected:
+
+  /**
+   * Obtain a writeable reference to a single parameter given a InputParameter name
+   */
+  template<typename T>
+  T & getControlParam(const std::string & name);
+
+  /**
+   * Obtain a writeable reference to a single parameter given a specific name
+   */
+  template<typename T>
+  T & getControlParamByName(const std::string & name);
+
+  /**
+   * Obtain pointers to all parameters given a InputParameter name
+   */
+  template<typename T>
+  std::vector<T*> getControlParamVector(const std::string & name);
+
+  /**
+   * Obtain pointers to all parameters given a specific name
+   */
+  template<typename T>
+  std::vector<T*> getControlParamVectorByName(const std::string & name);
+
+  /**
+   * Destructor
+   */
+  virtual ~ControlInterface(){}
+
+private:
+
+  /**
+   * A private constructor to limit the use of this interface to Control related objects
+   */
+  ControlInterface(const InputParameters & parameters);
+
+  /**
+   * Helper method for extracting iterators to InputParameter objects containing a property with the given name
+   * @param name The complete name of the desired parameter (only used for error messages)
+   * @param desired The desired parameter separated into components via tokenizeName method of this class
+   *
+   * Using the iterator allows for the public get methods to provide better error messages because the
+   * full object name can be reported.
+   *
+   * @see tokenizeName
+   */
+  template<typename T>
+  std::vector<InputParameterIterator> getValidInputParameterIteratorsByName(const std::string & name, const std::deque<std::string> & desired);
+
+  /**
+   * Helper method separating parameter name into system/group/param format
+   *
+   * [System]
+   *   [./group]
+   *     param = ...
+   *
+   */
+  std::deque<std::string> tokenizeName(const std::string & name);
+
+  /// Reference to the Control objects parameters
+  const InputParameters & _ci_parameters;
+
+  /// A Reference to MooseApp (needed to gain access to InputParameterWarehouse)
+  MooseApp & _ci_app;
+
+  /// A reference to the InputParameterWarehouse which is used for access the parameter objects
+  InputParameterWarehouse & _input_parameter_warehouse;
+
+  /// The object thread id, needed for access the warehouse
+  THREAD_ID _tid;
```

Comment:
I think I'm with @friedmud here. Also, because I am concerned about threaded behavior. We can make a choice on whether to thread the CL system or not. Personally I think it should _not_ be threaded because there's no tasks to distribute threads over so it's just overhead. However by not threading you have to think about how to interact with all of the controllable systems. You need to loop over all the thread copies of the InputParameter objects and make the same adjustement to all of them. The user would make the changes to the zeroth copy of the InputParameters and you'd have to apply that change to all the other copies.

As far as when/how to do all this? I'm fine with you getting something going that doesn't work with threads as long as you throw an error when trying to run with threads. If this is too much for a first take that's a compromise we can make.